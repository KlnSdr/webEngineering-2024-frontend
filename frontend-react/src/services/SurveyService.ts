import {CreateSurvey, Survey} from "../types/Surveys";;

class SurveyService {
    private static backendURL: string =
        process.env.REACT_APP_BACKEND_URL || "http://localhost:13000";

    public static getSurveysByUserId(userId: string): Promise<any[]> {
        return new Promise((resolve, reject) => {
            fetch(`${this.backendURL}/surveys/user/${userId}`).then((response: Response) => {
                if (!response.ok) {
                    throw new Error("Failed to load surveys.");
                }
                return response.json();
            }).then((data: Survey[]) => {
                resolve(data);
            }).catch((reason: any) => {
                reject(reason);
            });
        });
    }

    public static getSurveyById(surveyId: string | undefined): Promise<Survey> {
        return new Promise((resolve, reject) => {
            fetch(`${this.backendURL}/surveys/${surveyId}`, {credentials: "include"}).then((response: Response) => {
                if (!response.ok) {
                    throw new Error("Failed to load survey.");
                }
                return response.json();
            }).then((data: Survey) => {
                console.log(data);
                resolve(data);
            }).catch((reason: any) => {
                reject(reason);
            });
        });
    }
    public static voteSurvey(surveyId: number, recipeId: number): Promise<Survey>{
        return new Promise((resolve, reject) =>{
            fetch(`${this.backendURL}/surveys/${surveyId}/vote/${recipeId}`, {
                method: "PUT",
                credentials: "include",
                headers: {
                    "Content-Type": "application/json",
                },
            }).then((response: Response) => {
                if (!response.ok) {
                    throw new Error("Failed to vote.");
                }
                return response.json();
            }).then((data: Survey) => {
                resolve(data);
            }).catch((reason: any) => {
                reject(reason);
            });
        });
    }

    public static createSurvey(survey: CreateSurvey): Promise<void> {
        const surveyToSend = {
            title: survey.title,
            options: survey.options.map((recipe: any) => `/recipes/${recipe.id}`),
        };

        return new Promise((resolve, reject) => {
            fetch(`${this.backendURL}/surveys`, {
                method: "POST",
                credentials: "include",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(surveyToSend),
            }).then((response: Response) => {
                if (!response.ok) {
                    throw new Error("Could not save survey.");
                }
                return response.json();
            }).then((data: any) => {
                console.log(data);
                resolve();
            }).catch((reason: any) => {
                reject(reason);
            });
        });
    }
}

export default SurveyService;