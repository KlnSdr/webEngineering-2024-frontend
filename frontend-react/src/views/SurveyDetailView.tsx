import "../style/SurveyView.css";
import React, {useEffect, useState} from "react";
import ImageArea from "../components/ImageArea";
import {RecipeService} from "../services/RecipeService";
import {Recipe} from "../types/Recipes";
import MyRecipeBar from "../components/MyRecipeBar";
import Stack from "react-bootstrap/Stack";
import {Heading2} from "../components/Heading";
import UncheckCheckbox from "../components/UncheckCheckbox";
import {Survey} from "../types/Surveys";
import SurveyService from "../services/SurveyService";
import {useParams} from "react-router-dom";
import {User} from "../types/Users";
import {UserService} from "../services/UserService";


function SurveyDetailView() {

    const [recipes, setRecipe] = useState<Recipe[]>([]);
    const [creator, setCreator] = useState<User | null>(null);
    const [survey, setSurvey] = useState<Survey | undefined>(undefined);
    const {id} = useParams<{id: string}>();

    useEffect(() => {
        SurveyService.getSurveyById(id).then((survey) => {
            if (survey) {
                console.log("Survey data:", survey);
                setSurvey(survey);
                recipeSet(survey);
                UserService.getUser(survey.creator)
                    .then(setCreator)
                    .catch(_ => {
                        setCreator({
                            userName: "could not load username",
                            userId: -1
                        });
                });
            }
        }).catch ((error) => {
            console.error("Failed", error);
        });
    }, [id]);

    function recipeSet(survey: Survey) {
        let recipeIds = survey.options.map(option => option.replace("/recipes/", ""));
        Promise.all(recipeIds.map(id => RecipeService.getRecipeById(id)))
            .then(fetchedRecipes => {
                setRecipe(fetchedRecipes);
            })
            .catch(error => {
                console.error("Failed to fetch recipes", error);
            });
    }

    function handleVote( surveyId : number, recipeUri: number){
    SurveyService.voteSurvey(surveyId, recipeUri).then((survey) => {
        setSurvey(survey);
        });
    }

    function getVoteCount(recipeId: number): number {
        if (survey && survey.recipeVote) {
            return survey.recipeVote[`/recipes/${recipeId}`]|| 0;
        }
        return 0;
    }

    function formatDate(date: Date): string {
        let time = new Date(date);
        const day = String(time.getDate()).padStart(2, '0');
        const month = String(time.getMonth() + 1).padStart(2, '0'); // Months are zero-based
        const year = time.getFullYear();
        return `${day}.${month}.${year}`;
    }

if (!survey) {
    return <div>
        <Heading2 headingText={"Survey"}/>
        <h1>Survey not found</h1>
    </div>
}
    return (
        <div>
        <Heading2 headingText={"Survey"}/>
            <div>
                <h1>{survey.title}</h1>
                Erstellt von {creator?.userName || ""} am {formatDate(survey.creationDate)}
            </div>

            <div className="container">
                {recipes.map((recipe, index) => (
                    <div  key={index} className="container">
                    <div className="row mt-3">
                        <Stack direction={"horizontal"}>
                            <ImageArea origin={recipe.imgUri}/>
                            <MyRecipeBar Recipe={recipe}/>
                            {survey && (<UncheckCheckbox voteChange= {()=>handleVote(survey.id, recipe.id)} />)}
                            <div className="align-content-center "> Stimmen ---{'>'}  {getVoteCount(recipe.id)}</div>
                        </Stack>
                    </div>
                </div>
                ))}
            </div>
        </div>
    );
}

export default SurveyDetailView;
