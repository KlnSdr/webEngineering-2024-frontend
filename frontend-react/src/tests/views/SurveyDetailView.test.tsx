import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter} from "react-router-dom";
import SurveyDetailView from "../../views/SurveyDetailView";
import SurveyService from "../../services/SurveyService";

jest.mock("../../services/SurveyService");

const mockSurveys = SurveyService.getSurveyById as jest.Mock;
const mockSurveyTest = [{id: 1, title: "Test Survey", participants: ["user1" , "user2"],
    creator: "user1", recipeVote: {"recipe1" : 1},
    options: ["recipe1", "recipe2"], creationDate: new Date() }];

describe("SurveyDetailView Component", () => {

    test('matches snapshot for SurveyDetailView component if user is logged in', () => {

        (mockSurveys as jest.Mock).mockResolvedValueOnce({
            ok: true,
            json: async () => mockSurveyTest,
        });

        const { asFragment } = render(
            <MemoryRouter initialEntries={["/survey/view/:id"]}>
                <SurveyDetailView />
            </MemoryRouter>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test('matches snapshot for SurveyDetailView component if user is not logged in', () => {

        const mockError = new Error("Failed to load survey detail view");
        (mockSurveys as jest.Mock).mockRejectedValueOnce(mockError);
            const {asFragment} = render(
                <MemoryRouter initialEntries={["/survey/view/:id"]}>
                    <SurveyDetailView/>
                </MemoryRouter>
            );

            expect(asFragment()).toMatchSnapshot();
    });

});
