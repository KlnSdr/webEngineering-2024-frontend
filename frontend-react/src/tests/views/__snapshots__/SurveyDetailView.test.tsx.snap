// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurveyDetailView Component matches snapshot for SurveyDetailView component if user is logged in 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="FoodTitle"
    >
      Survey
    </h2>
    <h1>
      Survey not found
    </h1>
  </div>
</DocumentFragment>
`;

exports[`SurveyDetailView Component matches snapshot for SurveyDetailView component if user is not logged in 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="FoodTitle"
    >
      Survey
    </h2>
    <h1>
      Survey not found
    </h1>
  </div>
</DocumentFragment>
`;
